File: src/main/scala/us/zoom/data/dfence/providers/snowflake/revoke/SnowflakeRevokeGrantsCompiler.scala
package us.zoom.data.dfence.providers.snowflake.revoke

import us.zoom.data.dfence.playbook.model.PlaybookPrivilegeGrant
import us.zoom.data.dfence.providers.snowflake.grant.builder.SnowflakeGrantBuilder
import us.zoom.data.dfence.providers.snowflake.revoke.evaluator.GrantRevocationEvaluator
import us.zoom.data.dfence.providers.snowflake.revoke.index.PlaybookGrantIndexBuilder

import java.util.{List as JList, Map as JMap}
import scala.collection.parallel.CollectionConverters.*
import scala.collection.parallel.mutable.ParSeq
import scala.jdk.CollectionConverters.*

class SnowflakeRevokeGrantsCompiler:

  def compileRevokeGrants(
    privilegeGrants: JList[PlaybookPrivilegeGrant],
    currentGrantBuilders: JMap[String, SnowflakeGrantBuilder]
  ): JList[SnowflakeGrantBuilder] =
    val playbookGrantsIndex =
      PlaybookGrantIndexBuilder.buildPrivilegeGrantIndex(privilegeGrants.asScala)
    val evaluator           = GrantRevocationEvaluator(playbookGrantsIndex)

    currentGrantBuilders
      .values()
      .asScala
      .par
      .to(ParSeq)
      .filter(grantBuilder => evaluator.needsRevoke(grantBuilder.getGrant))
      .toList
      .sortBy(builder => builder.getKey)
      .asJava

File: src/main/java/us/zoom/data/dfence/providers/snowflake/revoke/SnowflakeRevokeGrantsCompiler.java
package us.zoom.data.dfence.providers.snowflake.revoke;

import io.vavr.control.Try;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import us.zoom.data.dfence.exception.RbacDataError;
import us.zoom.data.dfence.playbook.model.PlaybookPrivilegeGrant;
import us.zoom.data.dfence.providers.snowflake.grant.builder.SnowflakeGrantBuilder;
import us.zoom.data.dfence.providers.snowflake.revoke.evaluator.GrantRevocationEvaluator;
import us.zoom.data.dfence.providers.snowflake.revoke.index.PolicyGrantHashIndexer;

@Slf4j
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class SnowflakeRevokeGrantsCompiler {

  /**
   * Compiles grants to revoke by comparing current grants with playbook grants. Returns grants that
   * don't match any playbook grant.
   */
  public static List<SnowflakeGrantBuilder> compileRevokeGrants(
      List<PlaybookPrivilegeGrant> playbookGrants,
      Map<String, SnowflakeGrantBuilder> currentGrantBuilders) {
    return Try.of(
            () -> {
              GrantRevocationEvaluator evaluator =
                  new GrantRevocationEvaluator(PolicyGrantHashIndexer.create(playbookGrants));

              return currentGrantBuilders.values().stream()
                  .filter(grantBuilder -> evaluator.needsRevoke(grantBuilder.getGrant()))
                  .sorted(Comparator.comparing(SnowflakeGrantBuilder::getKey))
                  .collect(Collectors.toList());
            })
        .getOrElseThrow(
            e -> {
              log.error(
                  "Unexpected error compiling revoke grants: {} playbook grants provided",
                  playbookGrants.size(),
                  e);
              return new RbacDataError("Failed to compile revoke grants: " + e.getMessage(), e);
            });
  }
}
